File: stream/ResponseCollector.ts
/*
 * Copyright (C) 2021 Vaticle
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import {ErrorMessage} from "../common/errors/ErrorMessage";
import {TypeDBClientError} from "../common/errors/TypeDBClientError";
import {BlockingQueue} from "../common/util/BlockingQueue";

export class ResponseCollector<T> {

    private readonly _response_queues: { [requestId: string]: ResponseCollector.ResponseQueue<T> };

    constructor() {
        this._response_queues = {};
    }

    queue(uuid: string): ResponseCollector.ResponseQueue<T> {
        const queue = new ResponseCollector.ResponseQueue<T>();
        this._response_queues[uuid] = queue;
        return queue;
    }

    get(uuid: string) {
        return this._response_queues[uuid];
    }

    remove(requestId: string) {
        delete this._response_queues[requestId];
    }

    close(error?: Error | string) {
        Object.values(this._response_queues).forEach(collector => collector.close(error));
    }
}

export namespace ResponseCollector {

    import TRANSACTION_CLOSED = ErrorMessage.Client.TRANSACTION_CLOSED;
    import ILLEGAL_STATE = ErrorMessage.Internal.ILLEGAL_STATE;

    export class ResponseQueue<T> {
        private readonly _queue: BlockingQueue<QueueElement>;

        constructor() {
            this._queue = new BlockingQueue<QueueElement>()
        }

        async take(): Promise<T> {
            const element = await this._queue.take();
            if (element.isValue()) return (element as Value<T>).value;
            else if (element.isDone()) {
                if ((element as Done).hasError()) {
                    throw new TypeDBClientError((element as Done).error);
                } else {
                    throw new TypeDBClientError(TRANSACTION_CLOSED);
                }
            } else throw new TypeDBClientError(ILLEGAL_STATE);
        }

        put(element: T): void {
            this._queue.add(new Value(element));
        }

        close(error?: Error | string): void {
            this._queue.add(new Done(error));
        }
    }

    class QueueElement {

        isValue(): boolean {
            return false;
        }

        isDone(): boolean {
            return false;
        }

    }

    class Value<R> extends QueueElement {

        private readonly _value: R;

        constructor(value: R) {
            super();
            this._value = value;
        }

        get value(): R {
            return this._value;
        }

        isValue(): boolean {
            return true;
        }
    }

    class Done extends QueueElement {
        private readonly _error?: Error | string;

        constructor(error?: Error | string) {
            super();
            this._error = error;
        }

        hasError(): boolean {
            return this._error != null;
        }

        get error(): Error | string {
            return this._error;
        }

        isDone(): boolean {
            return true;
        }
    }
}

File: stream/ResponseCollector.java
/*
 * Copyright (C) 2021 Vaticle
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package com.vaticle.typedb.client.stream;

import com.vaticle.typedb.client.common.exception.TypeDBClientException;
import com.vaticle.typedb.common.collection.Either;
import io.grpc.StatusRuntimeException;

import javax.annotation.Nullable;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.UUID;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;
import java.util.concurrent.LinkedTransferQueue;

import static com.vaticle.typedb.client.common.exception.ErrorMessage.Client.TRANSACTION_CLOSED;
import static com.vaticle.typedb.client.common.exception.ErrorMessage.Client.TRANSACTION_CLOSED_WITH_ERRORS;
import static com.vaticle.typedb.client.common.exception.ErrorMessage.Internal.UNEXPECTED_INTERRUPTION;

public class ResponseCollector<R> {

    private final ConcurrentMap<UUID, Queue<R>> collectors;

    public ResponseCollector() {
        collectors = new ConcurrentHashMap<>();
    }

    synchronized Queue<R> queue(UUID requestId) {
        Queue<R> collector = new Queue<>();
        collectors.put(requestId, collector);
        return collector;
    }

    Queue<R> get(UUID requestId) {
        return collectors.get(requestId);
    }

    void remove(UUID requestID) {
        this.collectors.remove(requestID);
    }

    synchronized void close(@Nullable StatusRuntimeException error) {
        collectors.values().forEach(collector -> collector.close(error));
    }

    List<StatusRuntimeException> getErrors() {
        List<StatusRuntimeException> errors = new ArrayList<>();
        collectors.values().forEach(collector -> collector.getError().ifPresent(errors::add));
        return errors;
    }

    public static class Queue<R> {

        private final BlockingQueue<Either<Response<R>, Done>> responseQueue;
        private StatusRuntimeException error;

        Queue() {
            // TODO: switch LinkedTransferQueue to LinkedBlockingQueue once issue #351 is fixed
            responseQueue = new LinkedTransferQueue<>();
            error = null;
        }

        public R take() {
            try {
                Either<Response<R>, Done> response = responseQueue.take();
                if (response.isFirst()) return response.first().message();
                else if (this.error != null) throw new TypeDBClientException(TRANSACTION_CLOSED_WITH_ERRORS, error);
                else throw new TypeDBClientException(TRANSACTION_CLOSED);
            } catch (InterruptedException e) {
                throw new TypeDBClientException(UNEXPECTED_INTERRUPTION);
            }
        }

        public Optional<StatusRuntimeException> getError() {
            return Optional.ofNullable(error);
        }

        public void put(R response) {
            responseQueue.add(Either.first(new Response<>(response)));
        }

        public void close(@Nullable StatusRuntimeException error) {
            this.error = error;
            responseQueue.add(Either.second(new Done()));
        }

        private static class Response<R> {

            @Nullable
            private final R value;

            private Response(@Nullable R value) {
                this.value = value;
            }

            @Nullable
            private R message() {
                return value;
            }
        }

        private static class Done {
            private Done() {
            }
        }
    }
}
